Common/Util/RateLimit/LeakyBucket.cs

```diff
@@ -0,0 +1,168 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Threading;
+
+namespace QuantConnect.Util.RateLimit
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="ITokenBucket"/> that implements the leaky bucket algorithm
+    /// See: https://en.wikipedia.org/wiki/Leaky_bucket
+    /// </summary>
+    public class LeakyBucket : ITokenBucket
+    {
+        private readonly object _sync = new object();
+
+        private long _available;
+        private readonly long _capacity;
+        private readonly ISleepStrategy _sleep;
+        private readonly IRefillStrategy _refill;
+        private readonly ITimeProvider _timeProvider;
+
+        /// <summary>
+        /// Gets the total number of currently available tokens for consumption
+        /// </summary>
+        public long AvailableTokens
+        {
+            // synchronized read w/ the modification of available tokens in TryConsume
+            get { lock (_sync) return _available; }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LeakyBucket"/> class.
+        /// This constructor initializes the bucket using the <see cref="ThreadSleepStrategy.Sleep"/> with a 1 millisecond
+        /// sleep to prevent being CPU intensive and uses the <see cref="FixedIntervalRefillStrategy"/> to refill bucket
+        /// tokens according to the <paramref name="refillAmount"/> and <paramref name="refillInterval"/> parameters.
+        /// </summary>
+        /// <param name="capacity">The maximum number of tokens this bucket can hold</param>
+        /// <param name="refillAmount">The number of tokens to add to the bucket each <paramref name="refillInterval"/></param>
+        /// <param name="refillInterval">The interval which after passing more tokens are added to the bucket</param>
+        public LeakyBucket(long capacity, long refillAmount, TimeSpan refillInterval)
+            : this(capacity, ThreadSleepStrategy.Sleeping(1),
+                new FixedIntervalRefillStrategy(RealTimeProvider.Instance, refillAmount, refillInterval)
```

Comment:
Not sure how will `RealTimeProvider.Instance` work for backtesting, shouldn't it be related to the algorithms Time instead?

Reply:
In this case, the resources we're rate limiting is CPU time itself, hence using the real time provider.